src/graph.py

```diff
@@ -752,6 +750,104 @@ def perform_GO_enrichment_on_clusters(nxgraph, clustering):
+
+def summarize_grid_search(results, mode, top_n=-1): 
+    """
+    Summarizes results of `_grid_randomization` into a matrix where each row is a gene 
+    and each column is a parameter run. If summarizing "membership", entries will be 0 or 1
+    indicating whether or not a node appeared in each experiment. If summarizing "robustness"
+    or "specificity", entries indicate robustness or specificity values for each experiment. 
+    Also, node attributes columns are added for plotting purposes.
+
+    Arguments:
+        results (list of tuples): Results of `_grid_randomization` (paramstring, forest, augmented forest)
+        mode (str): Reported values "membership", "robustness", "specificity"
+        top_n (int): Takes the top_n values of the summary dataframe. top_n=-1 sets no threshold
+
+    Returns:
+        pd.DataFrame: Columns correspond to each parameter experiment, indexed by nodes
+    """
+    
+    # Exclude any degenerate results
+    results = {paramstring: graphs for paramstring, graphs in results.items() if graphs["augmented_forest"].number_of_nodes() > 0}
+    
+    if mode == "membership": # Summarize single-run parameter search
+        dfs = [pd.Series(1, index=graphs["augmented_forest"].nodes(), name=paramstring) for paramstring, graphs in results.items()]
+    elif mode == "robustness": # Summarize randomized robustness
+        dfs = [get_networkx_graph_as_dataframe_of_nodes(graphs["augmented_forest"])["robustness"].rename(paramstring) for paramstring, graphs in results.items()]
+    elif mode == "specificity": # Summarize randomized specificity
+        dfs = [get_networkx_graph_as_dataframe_of_nodes(graphs["augmented_forest"])["specificity"].rename(paramstring) for paramstring, graphs in results.items()]
+    else:
+        logger.info("`mode` must be one of the following: 'membership', 'robustness', or 'specificity'.")
+        return
+    
+    df = pd.concat(dfs, axis=1).fillna(0)
+    
+    # df can get quite large with many sparse entries, so let's filter for the top_n entries
+    if top_n == -1: return df
+    
+    if len(df) > top_n: 
+        cutoff = sorted(df.sum(axis=1).tolist(), reverse=True)[top_n]
+        df = df[df.sum(axis=1) > cutoff]
+    
+    return df
+
+
+def get_robust_subgraph_from_randomizations(nxgraph, max_size=400, min_component_size=5): 
+    """
+    Given a graph with robustness attributes, take the top `max_size` robust nodes and
+    prune any "small" components. 
+
+    Arguments:
+        nxgraph (networkx.Graph): Network from randomization experiment
+        max_size (int): Max size of robust network
+
+    Returns: 
+        networkx.Graph: Robust network
+
+    TODO: Potential alternative approaches
+    Approach 1: from entire network, attempt to remove lowest robustness node. If removal results in a component 
+    of size less than min_size, do not remove. 
+    Approach 2: select top max_size nodes based on robustness, then return subgraph. 
```

Comment:
Aren't you doing approach 2? 

Either way, this shouldn't be in the docstring -- leave it as a comment towards the top of the function.

Reply:
222c52d31c3f88a8db78a989c3cb9002795f3619